src/rgw/rgw_op.cc

```diff
@@ -6329,14 +6529,32 @@ void RGWCompleteMultipart::execute(optional_yield y)
     return;
   }
 
-
   if ((int)parts->parts.size() >
       s->cct->_conf->rgw_multipart_part_upload_limit) {
     op_ret = -ERANGE;
     return;
   }
 
   upload = s->bucket->get_multipart_upload(s->object->get_name(), upload_id);
+  ldpp_dout(this, 16) <<
+    fmt::format("INFO: {}->get_multipart_upload for obj {}, {} cksum_type {}",
+		s->bucket->get_name(),
+		s->object->get_name(), upload_id,
+		(!!upload) ? to_string(upload->cksum_type) : 0)
+		<< dendl;
+
+  rgw_placement_rule* dest_placement;
+  op_ret = upload->get_info(this, s->yield, &dest_placement);
+  if (op_ret < 0) {
+    /* XXX this fails consistently when !checksum */
+    ldpp_dout(this, 0) <<
+      "WARNING: MultipartUpload::get_info() for placement failed "
+		       << "ret=" << op_ret << dendl;
+    if (upload->cksum_type != rgw::cksum::Type::none) {
+      op_ret =  -ERR_INTERNAL_ERROR;
```

Comment:
ok, i guess we need this for `upload->cksum_type` below? we do need to know that so we can reject the request before calling `upload->complete()`. does `upload->get_info()` cache everything we'd need later in `RadosMultipartUpload::complete()`? if so i hope we could remove its own call to `RadosMultipartUpload::get_info()`

Reply:
I don't know.  I do not want to wait to merge checksums based on a theory that we should change any of the internals of the RadosMultipartUpload

Reply:
I will experiment with this tonight;  if it works as expected, we don't need to call get_info twice (but I'm not sure that adds actual overhead)

Reply:
`RadosMultipartUpload::get_info()` calls `read_op->prepare()` and `read_op->read()`. because of `set_prefetch_data()`, the `read()` call reads from the prefetched buffer instead of going back to the osd

if a repeated call to `RadosMultipartUpload::get_info()` used the same `sal::Object` for meta_obj, then both `prepare()` and `read()` could be satisfied from its `RGWObjectCtx` cache

however, `get_info()` uses `get_meta_obj()` to create its own local `sal::Object` handle. because they don't share a `RGWObjectCtx`, i expect that duplicated calls do add an extra round trip to the osd

Reply:
I believed that the call was required prior to invoking list_parts(), which is required to validate the checksums in the upload

Reply:
we also consult the checksum type of the original upload, but I think that's more a consistency check

Reply:
ok, so per webex discussion, our strategy to address this is to take daniel's change to cache attributes in RadosMultipartUpload.   thank you!